feat(docs): add telemetry overview (#263)
canonical/observability-stack: docs/explanation/telemetry/telemetry-overview.md

Reviewer: Your formatting is inconsistent with the rest of the Observability docs (other does don't have line limits)

---
myst:
  html_meta:
    description: "A short overview of how COS handles telemetry — collection, labeling, and correlation of logs, metrics, and traces."
---

(telemetry-overview)=

# Telemetry overview

Modern distributed systems generate three complementary signal types — **metrics**,

> **logs**, and **traces** — each answering a different question:

Reviewer: ```suggestion
Modern distributed systems generate three complementary signal types: metrics, logs, and traces. Each answer a different question:
```
Breaking it apart for easier reading

> | Signal  | Question it answers             | Example                                    |

Reviewer: IMO "Purpose" or "Definition" would be more meaningful than "Question it answers". If you change this though, it would affect the previous sentence (e.g., "each provide a different view of the system"). Purposes could be
Metrics: Show system state
Logs: Record events
Traces: Shows data flow across services

(or something like that)

This is more of a preference though - the question framing is still fine.

Author: Fixed in next commit.

|---------|---------------------------------|--------------------------------------------|
| Metrics | *How is the system behaving?*   | CPU usage, request latency percentiles      |
| Logs    | *What happened?*                | An error message with a stack trace         |

> | Traces  | *Where did the time go?*        | A request's journey across seven services   |

Reviewer: ```suggestion
| Traces  | *Where did the time go?*        | A request's journey across different services   |
```
or "multiple"

I know it's an "example", so the number is fine, but being so specific about that can distract from the intent of the example

COS collects all three and ties them together so you can jump from a

> spike on a dashboard straight to the log line or trace that explains it.

Reviewer: ```suggestion
COS collects all three types and stores them in separate backends. When signals share identifiers (such as labels or trace IDs), they can be correlated across systems. For example, linking a metric spike to the logs or traces associated with the same request.
```
Or something like that (reword as necessary). Trying to make it sound a bit more technical/detailed, less marketing-speak

## Collection

Signals reach COS through two paths:

> - **Pull** — Prometheus (or Mimir) receive metrics from endpoints that charms

Reviewer: ```suggestion
- **Pull**: Prometheus (or Mimir) receive metrics from endpoints that charms
```
This would apply to all here, but I think the colon looks more standard formatting for docs

> - **Pull** — Prometheus (or Mimir) receive metrics from endpoints that charms

Author: Fixed in next commit.

  expose automatically via `juju relate`.
- **Push** — Logs and traces are forwarded to Loki and Tempo, typically through
  an OpenTelemetry (OTel) Collector that charms configure for you.

In both cases the transport is set up by a Juju relation; no manual endpoint
configuration is required.

> See [telemetry collection](telemetry-collection) for the full picture.

Reviewer: ```suggestion
See [telemetry collection](telemetry-collection) for more information.
```
To remove "the full picture" idiom

## Labels

Every signal carries **labels** — key-value pairs such as
`juju_model`, `juju_application`, and `juju_unit` — that identify
*where* it came from.  Charm libraries inject these labels automatically,
giving you a consistent lens to filter and group telemetry across the

> entire estate.

Reviewer: ```suggestion
Every signal carries labels, which are key-value pairs that identify where the signal came from. Examples of labels used by COS include: 

- `juju_model`
- `juju_application`
- `juju_unit`

These labels are added by charm libraries and provide a consistent way to filter and group data across your entire estate.
```
Reword as necessary - just making it a bit clearer to read

Author: fixed in next commit.

For OTLP-based workloads the same Juju topology is mapped to
OTel resource attributes, so push-based and pull-based data share an

> identical naming scheme.

Reviewer: ```suggestion
For OTLP-based workloads, the same Juju topology is mapped to OTel resource attributes, so push-based and pull-based data share an identical naming scheme. This allows metrics, logs, and traces to be queried using the same identifiers, regardless of how they were collected.
```
Reword or reject that if it's not accurate - I added this for a tighter summary

See [telemetry labels](telemetry-labels) and [OTLP labels](telemetry-otlp-topology-labels) for details.

## Correlation

> Labels alone let you filter; **correlation** lets you navigate.  COS links

Reviewer: ```suggestion
Correlation lets you navigate based on shared identifiers between signals. COS links
```
I removed the labels part because it doesn't feel necessary to this description of what correlation is. Something to keep in mind is that users aren't always reading the doc all the way through top-to-bottom - they might just jump to this section. Which isn't to suggest that you _shouldn't_ connect across sections - it's more of a "does this actually add value?" question

signals together in two ways:

> - **Trace → Log** — trace and span IDs embedded in log records let Grafana

Reviewer: Same here as I said before - I'd swap to colons instead of a dash

Reviewer: Also I usually avoid those little arrows because they're hard to see / tell the difference between those and a dash - you could use an angled bracket? (Trace > Log)?

This one is just a preference though, I could go either way

Author: Done.

  jump from a slow trace to the exact log lines it produced.
- **Trace → Metric** — Prometheus exemplars attach trace IDs to metric
  samples, connecting a latency spike to the trace that caused it.

The backend integrations (Tempo ↔ Loki, Tempo ↔ Mimir) are enabled by
default in the COS Terraform modules; you only need to ensure your
applications emit the right context.

> See [telemetry correlation](telemetry-correlation) for a deeper dive.

Reviewer: ```suggestion
See [telemetry correlation](telemetry-correlation) for more information.
```
